### ISSUE TITLE
More RC-friendly file format

### ISSUE THREAD

**Issue by otoomet** [OpenOrienteering/mapper#1290]
### Steps to reproduce
1. create a new file
2. add a few simple objects
3. save it as _.xmap_ file

### Actual behaviour
The file will be large, and apparently includes undo information, zoom level, view and such.  Could not find a description of the file format.

### Expected behaviour
Option to save only strictly _map data_ (and preserve the order and formatting of objects saved).  The undo/view information will interfere with RC systems, such as _git_ and create a lot of spurious differences even when the actual map is not changed.  

I would like to keep some of the maps under RC and keep the diffs as simple as possible.  Ideally there will also be a some sort of graphical diff-merge tool (see #741) on top of the text-diffs.

### Configuration
Mapper Version: 0.8.4
Operating System: ubuntu 16

**Comment by dg0yt** [comment]
- The unstable version already creates additional linebreaks in the regular, compact *.omap* files, for better use in version control.
- The 'non-map' data is another story. At least undo history can be cleaned manually before saving for version control, via Edit > Clear undo/redo history. This should mitigate the most serious sub-issue.
- Side effects of color or symbol changes are probably the biggest show-stopper for this type of version control at the moment.

**Comment by dg0yt** [comment]
FTR the request for XML format documentation is tracked in #182.

**Comment by otoomet** [comment]
Thanks.

> Side effects of color or symbol changes are probably the biggest show-stopper for this type of version control at the moment.

I am not sure I understand what this means... 

As a side note, #1097 asks pretty much the opposite.

**Comment by dg0yt** [comment]
When you add/remove a color, it will trigger changes in a number of symbols.
When you add/remove a symbol, or changes symbol order, it will trigger changes in a number of objects.

I don't see a conflict with #1097, because it will affect objects only when they are modified anyway. It addresses revision control at an object level, while `git` and Co. can only do revision control on a file level. The question is how to implement #1097 without adding too much extra data.

**Comment by jmacura** [comment]
For the record: I was trying to write a conversion script from .omap to SVG, but found the file structure not really friendly to 3rd party tools. It contains the "noise" (undos & redos, map view), weirdly named elements ("barrier") and, despite being verbose in general, the coordinates itself are not atomic and hard to parse...

**Comment by dg0yt** [comment]
@jmacura If you want to process the XML, you should better use the `.xmap` variant, due to its verbosity. The `.omap` variant is optimized for fast saving and loading, and for small size. That is the reason why the coordinates are not explicit in this variant. Note that SVG uses a similar approach.

The barrier elements are used when a file format change would cause serious trouble (crash, misrepresentation) with slighly older versions of the software. This is easy to handle when reading an XML stream, but could be more difficult with an XML tree.

"Noise" isn't a fair assessment outside of the revision control perspective, at least as long as users expect to continue their work were they left it.

**Comment by otoomet** [comment]
I think this kind of metadata (like open files, views, zoom settings) is typically stored in a separate file by IDE-s and such.  Not sure if this might apply to Mapper.

**Comment by dg0yt** [comment]
> The barrier elements are used when a file format change would cause serious trouble (crash, misrepresentation) with slighly older versions of the software. This is easy to handle when reading an XML stream, but could be more difficult with an XML tree.

At the moment, the `barrier` element wraps elements which cause problems in older versions of Mapper. I would propose to modify this is as follows:
- Add an optional attribute `action` to `<barrier>`.
- If the `action` attribute has the value `skip`, readers shall skip the *following* element if they don't meet the version requirements expressed in the `<barrier>`. 
- If there are multiple elements to be skipped, now wrapped by a single common `<barrier>`, each element gets an individual preceding `<barrier action="skip">`.

Once implemented in Mapper, this would be enough to safeguard current versions against misbehaviour with files written by future versions. Other software could  handle it as needed, including just ignoring it.

Implementing this change for map reading now, it may become the standard for writing in a later version.

**Comment by dg0yt** [comment]
> I think this kind of metadata (like open files, views, zoom settings) is typically stored in a separate file by IDE-s and such. Not sure if this might apply to Mapper.

My first thought was that Mapper is more like a document editor (e.g. Word, PowerPoint), less like an IDE. (How many map makers do know what an IDE is?) However, when looking at the collection of templates used with a map, the perspective changes. Still, I'm not convinced that yet another file is really desirable.

**Comment by dg0yt** [comment]
> The file will be large, and apparently includes undo information, zoom level, view and such. Could not find a description of the file format.

Saving undo information can be turned off in the settings.

**Comment by aberlol** [comment]
Saving the same map file (without changes) in OOM on Linux and Android (same version number 0.9.4) does not produce exactly the same .omap file content. It seems as some <object><tags>-child nodes are swapping places.

This adds unnecessary changes in the commit history. It would be nice to either:

1. Be able to remove/turn off <tags>
2. Stop re-ordering the <tags>

**Comment by dg0yt** [comment]
> It seems as some `object/tags`-child nodes are swapping places. This adds unnecessary changes in the commit history. It would be nice to not re-ordering the `tags`.

Indeed. This is a `QHash` structure, and we don't care enough about the order of the keys.

> The `template` `path`-attribute is changed, even though the `relpath` is the same. Is (the absolute) path really necessary to save when we have a working relative path?

Unsure. You are looking at Linux and Android which do have a common root directory `/`. But on Windows, there are drive letters e.g. `C:`. If you just save the map on a different drive, there is no relative path.
OTOH, I already try to avoid absolute paths in test maps, example maps and symbol sets. The absolute directory layout may give hints about operating system, user name, or map project which might not be meant to be disclosed in some cases.

**Comment by aberlol** [comment]
> > It seems as some `object/tags`-child nodes are swapping places. This adds unnecessary changes in the commit history. It would be nice to not re-ordering the `tags`.
> 
> Indeed. This is a `QHash` structure, and we don't care enough about the order of the keys.
> 

I see! Then a quick fix would be to order the tags alphabetically every time you save. I understand that the order is un-important to the application... but it would really improve the revision-controlability at a low cost.

(In my case I just removed all tags with regex from the file since I didn't really need them anyway.)

> > The `template` `path`-attribute is changed, even though the `relpath` is the same. Is (the absolute) path really necessary to save when we have a working relative path?
> 
> Unsure. You are looking at Linux and Android which do have a common root directory `/`. But on Windows, there are drive letters e.g. `C:`. If you just save the map on a different drive, there is no relative path.
> OTOH, I already try to avoid absolute paths in test maps, example maps and symbol sets. The absolute directory layout may give hints about operating system, user name, or map project which might not be meant to be disclosed in some cases.

True, I see your point.

I like the idea posted earlier by @otoomet about creating a metadata-file that contains user specific information (even template settings). This would be a great solution in my opinion.

**Comment by dg0yt** [comment]
> Then a quick fix would be to order the tags alphabetically every time you save.

Sorting everytime you save is too expensive. For now, I switched away from `QHash` to a flat list, assuming that a) the number of items is low, and b) searching for particular tags happens only occassionally (e.g. on import with symbol assignment).

The new data structure is unit-tested, but I would appreciate if it could be complemented by some real-world testing of yesterday's master release.

**Comment by FelixFrog** [comment]
Ok so I've done some testing with git, branching, merging. There are two major things that are holding this back:
- The `count` attribute creates lots of merge conflicts. It isn't a problem if you only commit on the same branch, but as soon as you switch and, for example, you add a node to an object, there are merge conflicts to solve that are just a `count="35"` that conflicts with a `count="36"`. So are they absolutely necessarry? I think that as of now it's not that hard to teach OOM to read files that do not contain those attrributes.
- Colors are a pain, since if you change the order of a color all of the `priority` attributes have to change. A quick solution would be to be able to set priorities/z-index in a range from 0 to 65535 in the GUI (as most softwares do) which would also give more flexibility. Another solution would be to add an `id` attribute to them, and, for example, when a color is deleted, make it appear as missing in the color menu so that the user can replace it easily. This would also be a quality of life improvement.

As a long term solution, I think that the idea of a multi-file format is not that bad. One such idea would be to store everything in a folder

```
map-folder
├── colors.xml
├── index.json
├── metadata.json
├── parts
│   ├── part1.xml
│   └── part2.xml
├── symbols.xml
└── templates
    ├── template1.jpg
    └── template2.png
```
(Qt has good json support and in our case it would drastically reduce the size as well as improve human readibility)
We could then gzip/bzip the folder into something like "Mymap.map" and teach OOM to open both the compressed folder and the uncompressed one from `index.json`, so that it still remains one single file for beginners and it would also reduce  size (team could just store the uncompressed folders). This surely opens lots of possibilities, such as storying templates for teamwork and it would be an amizing tool for teams and not just using git.
This is nothing more than an idea but I think it's worth considering, since it would could be an extremely powerful solution in a far away future.

**Comment by dg0yt** [comment]
> * The `count` attribute creates lots of merge conflicts. It isn't a problem if you only commit on the same branch, but as soon as you switch and, for example, you add a node to an object, there are merge conflicts to solve that are just a `count="35"` that conflicts with a `count="36"`. So are they absolutely necessarry? I think that as of now it's not that hard to teach OOM to read files that do not contain those attrributes.

This attribute is not mandatory. It is used to pre-allocate memory in an informed way. With regard to versioning, this information might be moved to a sidecar file which is not under version control. If the file is missing or not consistent with the actual map file, there would be only a one-time performance impact.

> * Colors are a pain, since if you change the order of a color all of the `priority` attributes have to change.

It is not just colors. There are multiple spots where the current index in a list is used to refer to elements, instead of using a proper ID.

---

I'm not sure if it will be really feasible to leave merging forked work only to `git` and `notepad`.

I'm not willing to rush into a new/more complex format quickly. There are also other options to explore (for humans or revision control: YAML; binary: CBOR).
However, these formats, including the current XML, seem to share the common idea of having a tree of objects, array, and values. So it should be feasible to explore this step by step. 

In any case, we must be aware of the special case of loading and saving under memory/deadline pressure on Android. Updating few small files (only when modified) might even help, in contrast to updating a single big file.

**Comment by dg0yt** [comment]
>  json support

What I continue to trip over with JSON is its lacking tolerance for trailing comma. This invalid:
~~~
{
  "A": "aa",
  "BB": "b",
}
~~~
Similar, when removing the whole `BB` line, the comma from the `A` line has to be removed, too. This is failing the goal of being more revision-control-friedly.

**Comment by jmacura** [comment]
> 
> 
> > json support
> 
> What I continue to trip over with JSON is its lacking tolerance for trailing comma. 

That is a known limitation of JSON. One of multiple. Option is to use some superset of JSON, like [JSON5](https://json5.org/) or [YAML](https://yaml.org/), which improves usability but notably reduces interoperability.
Yet I have a strict feeling that a regular developer shall not bother with the syntax format specificities, as the parser/reader should handle it.

**Comment by dg0yt** [comment]
> Yet I have a strict feeling that a regular developer shall not bother with the syntax format specificities, as the parser/reader should handle it.

However, the context of discussion is automagic merge in `git` which is not a JSON parser. That's why I want to record it here explicitly.
(I mentioned YAML in https://github.com/OpenOrienteering/mapper/issues/1290#issuecomment-808711002.)

**Comment by otoomet** [comment]
Why do you want to have json format, @jmacura ?  From git perspective, xml is good enough (given formatting does not change).  I agree that json is somewhat more human readable but I am not sure it matters in case of larger projects (like a map).

**Comment by jmacura** [comment]
> However, the context of discussion is automagic merge in `git` which is not a JSON parser. That's why I want to record it here explicitly.

Good point.

> Why do you want to have json format, @jmacura ?

Actually, I am not the one who came up with JSON in this thread.

**Comment by FelixFrog** [comment]
> Actually, I am not the one who came up with JSON in this thread.

Yeah I did mention JSON but it was no more than an idea:sweat_smile:.
The reality is that the files in which we would store the actual map parts don't need to be super human readable, just git friendly (and possibily fast to parse/storage efficient). Ideally, they could just be a csv table with an object per line (with an id and the coords, kind of what is stored in a .omap file apart from the id) if it wasn't for other small things for which we need a little bit of structure. This would give best speed, file size, ease in parsing (no need for external libs) and git compatibility, but would require an external "data" file for little things that do not fit in the csv table (pattern tags, text, etc...). I am not proposing any definitive solution here, just giving out ideas.

I actually proposed JSON (and for the matter YAML/TOML/XML) for other small things such as map metadata (undos, redos), an index file, symbols, colors, etc..., but that is definitely less important and depends a lot on Qt libraries avaliability, ease in parsing, git compatibility and is certainly one of the last things to worry about (if anyone is interested [this](https://blog.mbedded.ninja/programming/serialization-formats/a-comparison-of-serialization-formats/) is a good resource to read).

In general at this point I think that the most sensible thing to do is to undersand wether creating a completely new format is worth the effort or if the current .omap/.xmap format is flexible enough to give good git compatibility.

**Comment by dg0yt** [comment]
Thanks for clarification. IMO There are two separate goals with regard to revision control:
- Improve revision-control friendliness of the existing single-file format.
- Study the feasibility of revision-control-friendly multi-file format.

For the latter, I would suggest to start with a good research of existing open GIS formats, instead of re-inventing the wheel again. But probably most GIS simply follow a more centralized approach (central database), while `git` et al. are successful due to there distributed, decentralized approach.

**Comment by mfbehrens** [comment]
I created [omap-git-driver](https://codeberg.org/orienteering-maps/omap-git-driver)

This is basicly a custom git filter and merge driver that normalises the xml and does a semantic merge of the xml if merging is needed.

This git filter does:
- autoformat xml
- strips out current view and some other elements
- strips out the count attributes (except on coord tags)
- adds a `unique_id` as an omap tag to each element in order to track elements in merges
- sorts the elements in object tags (I guess this is just because i sometimes have to add a tag in the filter which is added in the wrong place -> hacky but works for now)

The merge driver does:
- Parse the xml merge everything semantically
- If it finds conflicts, add the conflicted elements to a new map part and let the user resolve it

Example usage can be found [here](https://codeberg.org/orienteering-maps/de-karlsruhe)

There are a lot of problems with this approach:
- Hard to set up
- Currently only tested on Linux (rather .bat setup does not work)
- filter does not alter the actual file but only the version that gets pushed. When you clone/checkout/reset the file you get the version on git. When you added objects yourself but only committed them oomapper does not know about the `unique_id`s of the new elements since they are not in your real file, only in the committed version. (solveable with a pre-commit hook)
- Hard to set up (mainly for non-technical people)

My main goal is to build some platform where people can openly share, version, colaboratively edit and archive their maps.
Since this is pretty much what git forges like GitHub, Gitlab or Forgejo do as well I thought it is probably not worth developing a gitlab/forgejo clone but just use it.

Unfortunately, I think it is not really possible to version control orienteering maps in git (currently).
If you have any ideas that I missed please open some issues on the repo.

Otherweise I think the best option would be to put some version control settings into oomapper which are stored in the `.omap` file.
These settings should probably do something similar like my git filter but:
- Since the settings are stored in the file no complicated setup is needed
- The filter issue gets fixed since the file is written to file immediately

If nobody has any hard opinions against this i will maybe try to write some proper proposal on this

**Comment by dg0yt** [comment]
> * autoformat xml

But are you aware of the more verbose xmap format which already is autoformatted? This is how we store symbol sets and examples in this repository.

**Comment by mfbehrens** [comment]
Thank you very much. I was actually not aware.

I also think a good solution could also be to have document based settings.
I would like to be able to disable undo/redo history writing for a single file that I want to archive.

To gain more stability with the symbols (not quite sure where I read that) I was thinking about simply using the IDs directly.
I just found out that they don't seem to be unique, but I still think this could be worth a discussion.
This would even (partly) keep the data untouched if you change the style.

Another point would be to add some unique identifier to each object. This would afaik lay the basis for some omap merging, since you can actually be sure that objects are the same thing. This could also be a great usage for the document based setting. When I want to version-control my map I turn it on, but the normal user is not affected in bigger filesizes or else.

Other interesting document settings:

- undo/redo
- print settings
- add IDs
- Add absolute/relative/both filepaths to templates
- Disable count attribute for coords/objects/items whatever (could also be removed entirely from xmap


When reading about the multi-file-format I was thining about a file with the map data and one with all the last view, undo/redo history and so on that can be gitignored.
I personally dont think that splitting the fileformat into multiple files actually helps with versioning.
It mainly makes it more technical and encourages commiting only half of a change, which could make the file invalid.
Otherwise, if used properly it somehow gives the user a similar control of what they want to save per repo, which is somehow not that far from my document settings idea.

**Comment by mfbehrens** [comment]
Hello again,

after working a little with what I did and rereading this full thread, I now see the benefit of a fileformat as proposed [here](https://github.com/OpenOrienteering/mapper/issues/1290#issuecomment-808702851) by @FelixFrog.

However, I also see @dg0yt 's point to do some research on gis version control.
I found the following projects:
- https://github.com/koordinates/kart
- https://geogig.org/

The problem I see, is that these projects are not as established as git with less documentation, fewer tutorials, less gui applications. For that reason software like this will more likely be abandoned.

Then we also don't have "real geodata". Just a few MB of coordinates and attributes, but no GBs of scattered data around the world, where indexing would be important.

For that reason I would still support the idea of just using git for version control. It has been around for 20+ years and will be around for a long time. It is stable mature and can basicly still do what we want. When we say we need semantic merging this can be implemented via a custom merge driver.
